### PR TITLE
regions plugin: improve loop playback logic

### LINKED ISSUE #1868: Cancel Region Play Loop Properly
Hi, thanks for making this fantastic repo!

I'm trying to build a system with which users can freely navigate the video and loop whatever intervals they want. The region plugin seems to work properly at first glance. However,  it seems there are some bugs with the region plugin, specifically associated with looping.

[Problem]
1. Once I start looping on a region, I would not be able to use the native cursor from the wavesufer to navigate the video (you might still be able to select timestamps on the waveform but when you start playing it will jump back to the looping region). The only way to jump to other parts of the video is to click on other regions. 

2. If I loop on multiple regions, the browser would start being overloaded and becomes extremely slow. This happens when I've done looping a region and try to loop on other regions without stop the video first.

[Reproduce]
You could reproduce all these problems with the annotation and video-annotation examples.

I've tried to solve it by myself but haven't succeeded yet. I doubt it's because there's not any function actually "cancel" the looping. Does anyone have an idea?

Thanks

### PR DESCRIPTION
**Pull request by cadavre**
### Short description of changes:

`regions` plugin has doubled loop playback logic.

**First** can be found here, binding `out` event:
```js
    /* Bind audio events. */
    bindInOut() {
        // ...

        /* Loop playback. */
        this.on('out', () => {
            if (this.loop) {
                this.wavesurfer.play(this.start);
            }
        });
    }
```
https://github.com/katspaugh/wavesurfer.js/blob/master/src/plugin/regions/region.js#L336

**Second** is in `playLoop()` method:
```js
    /**
     * Play the audio region in a loop.
     * @param {number} start Optional offset to start playing at
     * */
    playLoop(start) {
        const s = start || this.start;
        this.wavesurfer.play(s);
        this.once('out', () => {
            const realTime = this.wavesurfer.getCurrentTime();
            if (realTime >= this.start && realTime <= this.end) {
                return this.playLoop();
            }
        });
    }
```
https://github.com/katspaugh/wavesurfer.js/blob/master/src/plugin/regions/region.js#L151

**The problem** is that both "loops" cannot be canceled in any way. So instead of doubling the same logics about looping and to allow disabling the loop – I propose to remove logics from `playLoop()` but keep method itself it for BC, **but** depend on `this.loop: boolean` flag to handle looping via `bindInOut()`. This way we can add `setLoop(boolean)` method to control the looping and remove strange logic from `playLoop()` if.

### Breaking in the external API:

No breaking changes.

### Breaking changes in the internal API:

No breaking changes.

### Todos/Notes:

- [ ] Update documentation.

### Related Issues and other PRs:

fixes #1868

EDIT: Removed TODO from example.

### CODE REVIEW

File: src/plugin/regions/region.js [hunk 1]
```diff
@@ -149,16 +149,18 @@ export class Region {
...
-        const s = start || this.start;
-        this.wavesurfer.play(s);
-        this.once('out', () => {
-            const realTime = this.wavesurfer.getCurrentTime();
-            if (realTime >= this.start && realTime <= this.end) {
-                return this.playLoop();
-            }
-        });
+        this.loop = true;
+        this.play(start);
```
Comment (osheroff) [inline src/plugin/regions/region.js:153]: need to keep the `start || this.start` logic that was in the previous version of the function.

Comment (cadavre) [reply src/plugin/regions/region.js:153]: `this.play()` is already having it.

**Comment by osheroff** [comment]
Overall the logic here looks correct to me.  I went to go test this code, as it turns out, region playback is totally broken in master, needs https://github.com/katspaugh/wavesurfer.js/pull/1970 to fix it. 

I tested with #1970 and this pr together and the looping code appears to work just fine.

File: src/plugin/regions/region.js (hunk 1 above)
Comment (osheroff) [reply src/plugin/regions/region.js:153]: got it, sorry.

**Comment by marizuccara** [comment]
Hi, this new logic seems to break again the loop: 
- if a region is looping and you click outside, the cursor seems to go to the clicked position, but after that returns looping on the region. 
- Again, if click inside other region, it goes in loop from one region to another.
I think is that because with the previous controls, the loop was interrupted if the current time was not inside the region.

**Comment by cadavre** [comment]
@marizuccara please provide scenarios of what should happen to your cases. I'd like to improve looping so all cases could be covered.

**Comment by cadavre** [comment]
@thijstriemstra changelog updated. It looks like also previous merged PR had wrong entry – made for 4.0.0 instead of 4.0.1 – so I also fixed this.

@marizuccara I reused code from before committing, to update loop logic based on current playback time vs region start-end times. Could you please test it in your case?

**Comment by thijstriemstra** [comment]
This branch has merge conflict.

> changelog updated. It looks like also previous merged PR had wrong entry – made for 4.0.0 instead of 4.0.1 – so I also fixed this.

4.0.1 was released today: https://github.com/katspaugh/wavesurfer.js/commit/aaf433ecaf2f032a28a0419401ed56958f67a17e

**Comment by marizuccara** [comment]
> @thijstriemstra changelog updated. It looks like also previous merged PR had wrong entry – made for 4.0.0 instead of 4.0.1 – so I also fixed this.
> 
> @marizuccara I reused code from before committing, to update loop logic based on current playback time vs region start-end times. Could you please test it in your case?

Hi @cadavre , tested your code! I think that with this unified logic, and with your latest fix, the solution seems very nice!

**Comment by cadavre** [comment]
@thijstriemstra conflicts resolved. Merged.

**Pull request merged**